src/gt4py/next/iterator/type_inference.py

```diff
@@ -506,53 +542,55 @@ def visit_FunCall(
                 )
             if node.fun.id.endswith("domain"):
                 for arg in node.args:
-                    constraints.add(
+                    self.constraints.add(
                         (
                             Val(kind=Value(), dtype=NAMED_RANGE_DTYPE, size=Scalar()),
-                            self.visit(arg, constraints=constraints, symtypes=symtypes),
+                            self.visit(arg, **kwargs),
                         )
                     )
                 return Val(kind=Value(), dtype=DOMAIN_DTYPE, size=Scalar())
 
-        fun = self.visit(node.fun, constraints=constraints, symtypes=symtypes)
-        args = Tuple.from_elems(*self.visit(node.args, constraints=constraints, symtypes=symtypes))
+        fun = self.visit(node.fun, **kwargs)
+        args = Tuple.from_elems(*self.visit(node.args, **kwargs))
         ret = TypeVar.fresh()
-        constraints.add((fun, FunctionType(args=args, ret=ret)))
+        self.constraints.add((fun, FunctionType(args=args, ret=ret)))
         return ret
 
     def visit_FunctionDefinition(
         self,
         node: ir.FunctionDefinition,
-        constraints: set[tuple[Type, Type]],
-        symtypes: dict[str, Type],
-    ) -> FunctionDefinitionType:
-        if node.id in symtypes:
-            raise TypeError(f"Multiple definitions of symbol {node.id}")
-
-        fun = self.visit(
-            ir.Lambda(params=node.params, expr=node.expr),
-            constraints=constraints,
-            symtypes=symtypes,
+        **kwargs,
+    ) -> LetPolymorphic:
+        fun = ir.Lambda(params=node.params, expr=node.expr)
+
+        # Since functions defined in a function definition are let-polymorphic we don't want
+        # their parameters to inherit the constraints of the arguments in a call to them. A simple
+        # way to do this is to run the type inference on the function itself and reindex its type
+        # vars when referencing the function, i.e. in a `SymRef`.
```

Comment:
Thanks for adding this explanation here. I think it would help eve more to expand a bit the docstring of the `LetPolmorphic`...

Reply:
Ok, I understand. We can keep as it is in the current PR and think about improvements in following PRs.